port yasps combatlog changes

Latest snapshot seems to have broken functionality on input streams:

For example:
```
java -jar parser/target/stats-0.1.0.one-jar.jar < ../testfiles/1193091757.dem 
```
This works fine.

However:
```
cat ../testfiles/1193091757.dem | java -jar parser/target/stats-0.1.0.one-jar.jar 
```
and
```
wget -qO- http://replay111.valve.net/570/1375055798_10047699.dem.bz2 | bunzip2 | java -jar parser/target/stats-0.1.0.one-jar.jar 
```
give an IOException "illegal seek".

Fixed on most of the test files, but now getting:
```
Exception in thread "main"
java.lang.reflect.InvocationTargetException

        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)

        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:606)
        at com.simontuffs.onejar.Boot.run(Boot.java:340)
        at com.simontuffs.onejar.Boot.main(Boot.java:166)
Caused by: java.lang.RuntimeException: java.io.EOFException
        at skadistats.clarity.event.Event.raise(Event.java:25)
        at skadistats.clarity.processor.runner.AbstractRunner.runWith(AbstractRunner.java:59)
        at yasp.Main.run(Main.java:430)
        at yasp.Main.main(Main.java:438)
        ... 6 more
Caused by: java.io.EOFException
        at skadistats.clarity.source.InputStreamSource.readByte(InputStreamSource.java:38)
        at skadistats.clarity.source.Source.readVarInt32(Source.java:30)
        at skadistats.clarity.processor.reader.InputSourceProcessor.processSource(InputSourceProcessor.java:65)
        at java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:599)
        at skadistats.clarity.event.AbstractInvocationPoint.invoke(AbstractInvocationPoint.java:73)
        at skadistats.clarity.event.Event.raise(Event.java:23)
        ... 9 more
```
This is on an ARDM match, https://github.com/yasp-dota/testfiles/blob/master/1189263979_ardm.dem

This same error occurs on an invalid file (shouldn't the error be "not a valid replay file"?)
```
Exception in thread "main"
java.lang.reflect.InvocationTargetException


        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)

        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)

        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

        at java.lang.reflect.Method.invoke(Method.java:606)

        at com.simontuffs.onejar.Boot.run(Boot.java:340)

        at com.simontuffs.onejar.Boot.main(Boot.java:166)

Caused by: java.io.EOFException

        at skadistats.clarity.source.InputStreamSource.readBytes(InputStreamSource.java:49)

        at skadistats.clarity.source.Source.readBytes(Source.java:20)

        at skadistats.clarity.source.Source.ensureDemoHeader(Source.java:70)

        at skadistats.clarity.processor.runner.SimpleRunner.<init>(SimpleRunner.java:11)

        at yasp.Main.run(Main.java:430)

        at yasp.Main.main(Main.java:438)

        ... 6 more
```

--- a/src/main/java/skadistats/clarity/processor/gameevents/CombatLog.java
+++ b/src/main/java/skadistats/clarity/processor/gameevents/CombatLog.java
@@ -12,6 +12,7 @@ import skadistats.clarity.processor.stringtables.UsesStringTable;
 
 import java.util.LinkedList;
 import java.util.List;
+import java.util.Arrays;
 
 @Provides({OnCombatLogEntry.class})
 public class CombatLog {
@@ -20,44 +21,6 @@ public class CombatLog {
     public static final String GAME_EVENT_NAME   = "dota_combatlog";
 
     private final List<Entry> logEntries = new LinkedList<>();
-
-    @OnGameEventDescriptor(GAME_EVENT_NAME)
-    @UsesStringTable(CombatLog.STRING_TABLE_NAME)
-    public void onGameEventDescriptor(Context ctx, GameEventDescriptor descriptor){
-        typeIdx = descriptor.getIndexForKey("type");
-        sourceNameIdx = descriptor.getIndexForKey("sourcename");
-        targetNameIdx = descriptor.getIndexForKey("targetname");
-        attackerNameIdx = descriptor.getIndexForKey("attackername");
-        inflictorNameIdx = descriptor.getIndexForKey("inflictorname");
-        attackerIllusionIdx = descriptor.getIndexForKey("attackerillusion");
-        targetIllusionIdx = descriptor.getIndexForKey("targetillusion");
-        valueIdx = descriptor.getIndexForKey("value");
-        healthIdx = descriptor.getIndexForKey("health");
-        timestampIdx = descriptor.getIndexForKey("timestamp");
-        targetSourceNameIdx = descriptor.getIndexForKey("targetsourcename");
-
-        timestampRawIdx = descriptor.getIndexForKey("timestampraw");
-        attackerHeroIdx = descriptor.getIndexForKey("attackerhero");
-        targetHeroIdx = descriptor.getIndexForKey("targethero");
-        abilityToggleOnIdx = descriptor.getIndexForKey("ability_toggle_on");
-        abilityToggleOffIdx = descriptor.getIndexForKey("ability_toggle_off");
-        abilityLevelIdx = descriptor.getIndexForKey("ability_level");
-        goldReasonIdx = descriptor.getIndexForKey("gold_reason");
-    }
-
-    @OnGameEvent(GAME_EVENT_NAME)
-    public void onGameEvent(Context ctx, GameEvent gameEvent) {
-        logEntries.add(new Entry(ctx, gameEvent));
-    }
-
-    @OnTickEnd
-    public void onTickEnd(Context ctx, boolean synthetic) {
-        Event<OnCombatLogEntry> ev = ctx.createEvent(OnCombatLogEntry.class, Entry.class);
-        for (Entry e : logEntries) {
-            ev.raise(e);
-        }
-        logEntries.clear();
-    }
 
     private int typeIdx;
     private int sourceNameIdx;
@@ -77,8 +40,71 @@ public class CombatLog {
     private Integer abilityToggleOffIdx;
     private Integer abilityLevelIdx;
     private Integer goldReasonIdx;
+    private Integer xpReasonIdx;
+    //temporary manually set idx values for fields missing in gameeventdescriptor
+    Integer stunDurationIdx = 16;
+    Integer slowDurationIdx = 17;
+    Integer locationXIdx = 21;
+    Integer locationYIdx = 22;
+    Integer modifierDurationIdx = 25;
+    Integer lastHitsIdx = 27;
+    Integer attackerTeamIdx = 28;
+    Integer targetTeamIdx = 29;
+    Integer obsWardsPlacedIdx = 30;
+    Integer assist_player0Idx = 31;
+	Integer assist_player1Idx = 32;
+	Integer assist_player2Idx = 33;
+	Integer assist_player3Idx = 34;
+	Integer stack_countIdx = 35;
+	Integer hidden_modifierIdx = 36;
+	
+    @OnGameEventDescriptor(GAME_EVENT_NAME)
+    @UsesStringTable(CombatLog.STRING_TABLE_NAME)
+    public void onGameEventDescriptor(Context ctx, GameEventDescriptor descriptor){
+        typeIdx = descriptor.getIndexForKey("type");
+        sourceNameIdx = descriptor.getIndexForKey("sourcename");
+        targetNameIdx = descriptor.getIndexForKey("targetname");
+        attackerNameIdx = descriptor.getIndexForKey("attackername");
+        inflictorNameIdx = descriptor.getIndexForKey("inflictorname");
+        attackerIllusionIdx = descriptor.getIndexForKey("attackerillusion");
+        targetIllusionIdx = descriptor.getIndexForKey("targetillusion");
+        valueIdx = descriptor.getIndexForKey("value");
+        healthIdx = descriptor.getIndexForKey("health");
+        timestampIdx = descriptor.getIndexForKey("timestamp");
+        targetSourceNameIdx = descriptor.getIndexForKey("targetsourcename");
+        timestampRawIdx = descriptor.getIndexForKey("timestampraw");
+        attackerHeroIdx = descriptor.getIndexForKey("attackerhero");
+        targetHeroIdx = descriptor.getIndexForKey("targethero");
+        abilityToggleOnIdx = descriptor.getIndexForKey("ability_toggle_on");
+        abilityToggleOffIdx = descriptor.getIndexForKey("ability_toggle_off");
+        abilityLevelIdx = descriptor.getIndexForKey("ability_level");
+        goldReasonIdx = descriptor.getIndexForKey("gold_reason");
+        xpReasonIdx = descriptor.getIndexForKey("xp_reason");
+        //descriptor only contains a subset of the keys mentioned in protobufs
+        /*
+        stunDurationIdx = descriptor.getIndexForKey("stun_duration");
+        slowDurationIdx = descriptor.getIndexForKey("slow_duration");
+        locationXIdx = descriptor.getIndexForKey("location_x");
+        locationYIdx = descriptor.getIndexForKey("location_y");
+        */
+        //System.err.println(Arrays.toString(descriptor.getKeys()));
+    }
 
-    public class Entry {
+    @OnGameEvent(GAME_EVENT_NAME)
+    public void onGameEvent(Context ctx, GameEvent gameEvent) {
+        logEntries.add(new Entry(ctx, gameEvent));
+    }
+
+    @OnTickEnd
+    public void onTickEnd(Context ctx, boolean synthetic) {
+        Event<OnCombatLogEntry> ev = ctx.createEvent(OnCombatLogEntry.class, Entry.class);
+        for (Entry e : logEntries) {
+            ev.raise(e);
+        }
+        logEntries.clear();
+    }
+    
+     public class Entry {
 
         private final StringTable combatLogNames;
         private final GameEvent event;
@@ -92,11 +118,6 @@ public class CombatLog {
             return idx == 0 ? null : combatLogNames.getNameByIndex(idx);
         }
 
-        private String translate(String in) {
-            // TODO: translate modifier_XXX, or npc_hero_XXX into correct names...
-            return in;
-        }
-
         public GameEvent getGameEvent() {
             return event;
         }
@@ -106,27 +127,19 @@ public class CombatLog {
         }
 
         public String getSourceName() {
-            return translate(readCombatLogName((int)event.getProperty(sourceNameIdx)));
+            return readCombatLogName((int)event.getProperty(sourceNameIdx));
         }
 
         public String getTargetName() {
-            return translate(readCombatLogName((int)event.getProperty(targetNameIdx)));
-        }
-
-        public String getTargetNameCompiled() {
-            return getTargetName() + (isTargetIllusion() ? " (Illusion)" : "");
+            return readCombatLogName((int)event.getProperty(targetNameIdx));
         }
 
         public String getAttackerName() {
-            return translate(readCombatLogName((int)event.getProperty(attackerNameIdx)));
-        }
-
-        public String getAttackerNameCompiled() {
-            return getAttackerName() + (isAttackerIllusion() ? " (Illusion)" : "");
+            return readCombatLogName((int)event.getProperty(attackerNameIdx));
         }
 
         public String getInflictorName() {
-            return translate(readCombatLogName((int)event.getProperty(inflictorNameIdx)));
+            return readCombatLogName((int)event.getProperty(inflictorNameIdx));
         }
 
         public boolean isAttackerIllusion() {
@@ -150,19 +163,11 @@ public class CombatLog {
         }
 
         public String getTargetSourceName() {
-            return translate(readCombatLogName((int)event.getProperty(targetSourceNameIdx)));
+            return readCombatLogName((int)event.getProperty(targetSourceNameIdx));
         }
 
         public float getTimestampRaw() {
             return event.getProperty(timestampRawIdx);
-        }
-
-        public boolean isAttackerHero() {
-            return event.getProperty(attackerHeroIdx);
-        }
-
-        public boolean isTargetHero() {
-            return event.getProperty(targetHeroIdx);
         }
 
         public boolean isAbilityToggleOn() {
@@ -176,10 +181,66 @@ public class CombatLog {
         public int getAbilityLevel() {
             return event.getProperty(abilityLevelIdx);
         }
-
+        //TODO: for full safety, all getters should check for null on Idx
+        public String toString(){
+            //print the underlying gameevent
+            return event.toString();
+        }
+        public boolean isTargetHero() {
+            if (targetHeroIdx==null) {
+                return true;
+            }
+            return event.getProperty(targetHeroIdx);
+        }
+        public boolean isAttackerHero() {
+            if (attackerHeroIdx==null){
+                return true;
+            }
+            return event.getProperty(attackerHeroIdx);
+        }
         public int getGoldReason() {
+            if (goldReasonIdx==null){
+                return 0;
+            }
             return event.getProperty(goldReasonIdx);
         }
+        public int getXpReason() {
+            if (xpReasonIdx==null){
+                return 0;
+            }
+            return event.getProperty(xpReasonIdx);
+        }
+        public String getValueName(){
+            return readCombatLogName(getValue());
+        }
+        /*
+        public float getStunDuration() {
+            return event.getProperty(stunDurationIdx);
+        }
+        public float getSlowDuration() {
+            return event.getProperty(slowDurationIdx);
+        }
+        public int getLocationX() {
+            return event.getProperty(locationXIdx);
+        }
+        public int getLocationY() {
+            return event.getProperty(locationYIdx);
+        }
+        public float getModifierDuration(){
+            return event.getProperty(modifierDurationIdx);
+        }
+        public int getLastHits(){
+            return event.getProperty(lastHitsIdx);
+        }
+        public int getAttackerTeam(){
+            return event.getProperty(attackerTeamIdx);
+        }
+        public int getTargetTeam(){
+            return event.getProperty(targetTeamIdx);
+        }
+        public int getObsWardsPlaced(){
+            return event.getProperty(obsWardsPlacedIdx);
+        }
+        */
     }
-
 }

